finality-aleph/src/validator_network/manager/mod.rs

```diff
@@ -30,191 +41,165 @@ impl Display for SendError {
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(
-                f,
-                "outgoing only - {:?} [{}];",
-                self.outgoing_peers.len(),
-                peers
-            )?;
+        match wanted_incoming {
+            0 => write!(f, "not expecting any incoming connections; ")?,
+            _ => {
+                write!(f, "expecting {:?} incoming connections; ", wanted_incoming)?;
+                match self.incoming_peers.is_empty() {
+                    true => write!(f, "WARNING! No incoming peers even though we expected tham, maybe connecting to us is impossible; ")?,
+                    false => write!(
+                            f,
+                            "have - {:?} [{}]; ",
+                            self.incoming_peers.len(),
+                            pretty_authority_id_set(&self.incoming_peers),
+                    )?,
+                }
+                if !self.missing_incoming.is_empty() {
+                    write!(
+                        f,
+                        "missing - {:?} [{}]; ",
+                        self.missing_incoming.len(),
+                        pretty_authority_id_set(&self.missing_incoming),
+                    )?;
+                }
+            }
         }
 
-        if !self.missing_peers.is_empty() {
-            let peers = self
-                .missing_peers
-                .iter()
-                .map(|authority_id| authority_id.to_short_string())
-                .collect::<Vec<_>>()
-                .join(", ");
-            write!(f, "missing - {:?} [{}];", self.missing_peers.len(), peers)?;
+        match wanted_outgoing {
+            0 => write!(f, "not attempting any outgoing connections; ")?,
+            _ => {
+                write!(f, "attempting {:?} outgoing connections; ", wanted_outgoing)?;
+                if !self.outgoing_peers.is_empty() {
+                    write!(
+                        f,
+                        "have - {:?} [{}]; ",
+                        self.incoming_peers.len(),
+                        pretty_authority_id_set(&self.outgoing_peers),
+                    )?;
+                }
+                if !self.missing_outgoing.is_empty() {
+                    write!(
+                        f,
+                        "missing - {:?} [{}]; ",
+                        self.missing_incoming.len(),
+                        pretty_authority_id_set(&self.missing_outgoing),
+                    )?;
+                }
+            }
         }
 
         Ok(())
     }
 }
 
+/// Possible results of adding connections.
+#[derive(Debug, PartialEq, Eq)]
+pub enum AddResult {
+    /// We do not want to maintain a connection with this peer.
+    Uninterested,
+    /// Connection added.
+    Added,
+    /// Old connection replaced with new one.
+    Replaced,
+}
+
 impl<A: Data, D: Data> Manager<A, D> {
```

Comment:
weak suggestion: I would keep `impl Manager` closer to `struct Manager`, in particular, I wouldn't separate them with other big blocks

Reply:
Fair, I usually structure code in files in a way that makes more sense when read from beginning to end, and then the initial declaration of the struct serves as an introduction, but that's probably not really a good practice. Moved some of the code around to keep related stuff closer to each other.